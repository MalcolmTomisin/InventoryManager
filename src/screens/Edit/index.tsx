import React, {useState} from 'react';
import {ScrollView} from 'react-native';
import {TextInput, Button} from '../../components';
import styles from './style';
import {GlobalState} from '../../types';
import {useSelector, useDispatch} from 'react-redux';
import appActions from '../../provider/actions';
import utils from '../../utils';

const ACTIONS = {
  handleName: 1,
  handlePrice: 2,
  handleStock: 3,
  handleDescription: 4,
};

export default function AddItem({navigation: {navigate}}) {
  const {setOfNames, inventory} = useSelector((state: GlobalState) => state);
  const dispatch = useDispatch();
  const [form, handleForm] = useState({
    name: '',
    price: '',
    totalStock: '',
    description: '',
  });
  const [error, handleError] = useState({
    name: '',
    price: '',
    totalStock: '',
    description: '',
  });
  const [disabled, setDisabled] = useState<boolean>(true);
  const addItemToInventory = () => {
    setOfNames?.add(form.name.trim().toLowerCase());
    inventory.unshift({
      ...form,
      price: parseInt(form.price),
      totalStock: parseInt(form.totalStock),
    });
    dispatch(appActions.addItem({inventory, setOfNames}));
    navigate(utils.routes.HOME);
  };

  const checkAllInputs = () => {
    if (form.name.trim().length < 3) {
      return handleError({...error, name: 'Invalid name'});
    }
    if (setOfNames?.has(form.name.trim().toLowerCase())) {
      return handleError({...error, name: 'Name already taken'});
    }
    if (!form.price.match(/^\d+$/)) {
      return handleError({...error, price: 'invalid price'});
    }
    if (!form.totalStock.match(/^\d+$/)) {
      return handleError({...error, totalStock: 'Invalid stock'});
    }
    if (form.description.trim().split(' ').length < 3) {
      return handleError({...error, description: 'Add more description'});
    }
    return setDisabled(false);
  };

  const handleTextInput = (text: string, flag: number) => {
    switch (flag) {
      case ACTIONS.handleName:
        handleError({...error, name: ''});
        return handleForm({...form, name: text});
      case ACTIONS.handlePrice:
        handleError({...error, price: ''});
        return handleForm({...form, price: text});
      case ACTIONS.handleStock:
        handleError({...error, totalStock: ''});
        return handleForm({...form, totalStock: text});
      case ACTIONS.handleDescription:
        handleError({...error, description: ''});
        return handleForm({...form, description: text});
    }
  };
  return (
    <ScrollView style={styles.container}>
      <TextInput
        label="Name"
        placeholder="Tomisin Alu"
        containerStyle={styles.input}
        value={form.name}
        onChangeText={text => {
          handleTextInput(text, ACTIONS.handleName);
        }}
        error={error.name}
        onEndEditing={() => checkAllInputs()}
      />
      <TextInput
        label="Price"
        placeholder="â‚¦ 500"
        containerStyle={styles.input}
        keyboardType="numeric"
        value={form.price}
        onChangeText={text => {
          handleTextInput(text, ACTIONS.handlePrice);
        }}
        error={error.price}
        onEndEditing={() => checkAllInputs()}
      />
      <TextInput
        label="Total stock (Qty)"
        placeholder="65"
        containerStyle={styles.input}
        keyboardType="numeric"
        value={form.totalStock}
        onChangeText={text => {
          handleTextInput(text, ACTIONS.handleStock);
        }}
        error={error.totalStock}
        onEndEditing={() => checkAllInputs()}
      />
      <TextInput
        label="Description"
        placeholder="..."
        multiline
        containerStyle={styles.input}
        value={form.description}
        onChangeText={text => {
          handleTextInput(text, ACTIONS.handleDescription);
        }}
        error={error.description}
        onEndEditing={() => checkAllInputs()}
      />
      <Button
        label="Add"
        disabled={disabled}
        containerStyle={styles.input}
        onPress={addItemToInventory}
      />
    </ScrollView>
  );
}