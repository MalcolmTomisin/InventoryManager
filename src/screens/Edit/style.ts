import {StyleSheet} from 'react-native';

const styles = StyleSheet.create({
  container: {
    flex: 1,
    paddingStart: 15,
  },
  input: {
    marginVertical: 10,
  },
});

export default styles;
