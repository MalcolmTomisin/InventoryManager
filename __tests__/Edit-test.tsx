import 'react-native';
import React from 'react';
import Edit from '../src/screens/Edit';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

const createTestProps = (props: Object) => ({
  navigation: {
    state: {params: {}},
    dispatch: jest.fn(),
    goBack: jest.fn(),
    dismiss: jest.fn(),
    navigate: jest.fn(),
    openDrawer: jest.fn(),
    closeDrawer: jest.fn(),
    toggleDrawer: jest.fn(),
    getParam: jest.fn(),
    setParams: jest.fn(),
    addListener: jest.fn(),
    push: jest.fn(),
    replace: jest.fn(),
    pop: jest.fn(),
    popToTop: jest.fn(),
    isFocused: jest.fn(),
  },
  route: {
    params: {itemId: null},
  },
  ...props,
});

jest.mock('react-redux', () => {
  const ActualReactRedux = jest.requireActual('react-redux');
  return {
    ...ActualReactRedux,
    useSelector: jest.fn().mockImplementation(() => {
      return {inventory: [], setOfNames: new Set()};
    }),
    useDispatch: jest.fn().mockImplementation(() => {
      return () => {};
    }),
  };
});

let props = createTestProps({});

it('renders correctly', () => {
  const tree = renderer.create(<Edit {...props} />).toJSON();
  expect(tree).toMatchSnapshot();
});
