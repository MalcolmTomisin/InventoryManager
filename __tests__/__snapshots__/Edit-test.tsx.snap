// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<RCTScrollView
  contentContainerStyle={
    Object {
      "flex": 1,
      "paddingStart": 15,
    }
  }
>
  <View>
    <View
      style={
        Array [
          Object {
            "paddingVertical": 5,
            "width": 637.5,
          },
          Object {
            "marginVertical": 10,
          },
        ]
      }
    >
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Aileron-Bold",
            "fontSize": 16,
          }
        }
      >
        Name
      </Text>
      <View
        style={
          Array [
            Object {
              "borderBottomColor": "#9393AA",
            },
            Object {
              "borderBottomWidth": 1,
              "justifyContent": "center",
            },
          ]
        }
      >
        <TextInput
          allowFontScaling={true}
          onChangeText={[Function]}
          onEndEditing={[Function]}
          onFocus={[Function]}
          placeholder="Tomisin Alu"
          placeholderTextColor="#9393AA"
          rejectResponderTermination={true}
          style={
            Array [
              Object {
                "borderWidth": 0,
                "color": "#272755",
                "fontFamily": "Aileron-Regular",
                "fontSize": 14,
              },
              Object {
                "height": 48,
              },
            ]
          }
          underlineColorAndroid="transparent"
          value=""
        />
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "paddingVertical": 5,
            "width": 637.5,
          },
          Object {
            "marginVertical": 10,
          },
        ]
      }
    >
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Aileron-Bold",
            "fontSize": 16,
          }
        }
      >
        Price
      </Text>
      <View
        style={
          Array [
            Object {
              "borderBottomColor": "#9393AA",
            },
            Object {
              "borderBottomWidth": 1,
              "justifyContent": "center",
            },
          ]
        }
      >
        <TextInput
          allowFontScaling={true}
          keyboardType="numeric"
          onChangeText={[Function]}
          onEndEditing={[Function]}
          onFocus={[Function]}
          placeholder="₦ 500"
          placeholderTextColor="#9393AA"
          rejectResponderTermination={true}
          style={
            Array [
              Object {
                "borderWidth": 0,
                "color": "#272755",
                "fontFamily": "Aileron-Regular",
                "fontSize": 14,
              },
              Object {
                "height": 48,
              },
            ]
          }
          underlineColorAndroid="transparent"
          value=""
        />
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "paddingVertical": 5,
            "width": 637.5,
          },
          Object {
            "marginVertical": 10,
          },
        ]
      }
    >
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Aileron-Bold",
            "fontSize": 16,
          }
        }
      >
        Total stock (Qty)
      </Text>
      <View
        style={
          Array [
            Object {
              "borderBottomColor": "#9393AA",
            },
            Object {
              "borderBottomWidth": 1,
              "justifyContent": "center",
            },
          ]
        }
      >
        <TextInput
          allowFontScaling={true}
          keyboardType="numeric"
          onChangeText={[Function]}
          onEndEditing={[Function]}
          onFocus={[Function]}
          placeholder="65"
          placeholderTextColor="#9393AA"
          rejectResponderTermination={true}
          style={
            Array [
              Object {
                "borderWidth": 0,
                "color": "#272755",
                "fontFamily": "Aileron-Regular",
                "fontSize": 14,
              },
              Object {
                "height": 48,
              },
            ]
          }
          underlineColorAndroid="transparent"
          value=""
        />
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "paddingVertical": 5,
            "width": 637.5,
          },
          Object {
            "marginVertical": 10,
          },
        ]
      }
    >
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Aileron-Bold",
            "fontSize": 16,
          }
        }
      >
        Description
      </Text>
      <View
        style={
          Array [
            Object {
              "borderBottomColor": "#9393AA",
            },
            Object {
              "borderBottomWidth": 1,
              "justifyContent": "center",
            },
          ]
        }
      >
        <TextInput
          allowFontScaling={true}
          multiline={true}
          onChangeText={[Function]}
          onEndEditing={[Function]}
          onFocus={[Function]}
          placeholder="..."
          placeholderTextColor="#9393AA"
          rejectResponderTermination={true}
          style={
            Array [
              Object {
                "borderWidth": 0,
                "color": "#272755",
                "fontFamily": "Aileron-Regular",
                "fontSize": 14,
              },
              Object {
                "height": 80,
              },
            ]
          }
          underlineColorAndroid="transparent"
          value=""
        />
      </View>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#9393AA",
          "borderRadius": 5,
          "height": 48,
          "justifyContent": "center",
          "marginVertical": 10,
          "opacity": 1,
          "padding": 10,
          "width": 300,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#FFFFFF",
            "fontFamily": "GothamMedium",
            "fontSize": 12,
          }
        }
      >
        Update
      </Text>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#F74300",
          "borderRadius": 43.2,
          "bottom": 15,
          "height": 68,
          "justifyContent": "center",
          "opacity": 1,
          "position": "absolute",
          "right": 25,
          "width": 68,
        }
      }
      testID="delete"
    >
      <Text
        allowFontScaling={false}
        style={
          Array [
            Object {
              "color": "#fff",
              "fontSize": 24,
            },
            undefined,
            Object {
              "fontFamily": "FontAwesome",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        
      </Text>
    </View>
  </View>
</RCTScrollView>
`;
